phoenix-core/src/main/java/org/apache/phoenix/hbase/index/GlobalIndexer.java

```diff
@@ -0,0 +1,949 @@
+              metricSource.incrementSlowDuplicateKeyCheckCalls();
+          }
+          metricSource.updateDuplicateKeyCheckTime(duration);
+      }
+  }
+
+  @Override
+  public void preBatchMutate(ObserverContext<RegionCoprocessorEnvironment> c,
+      MiniBatchOperationInProgress<Mutation> miniBatchOp) throws IOException {
+      if (this.disabled) {
+          return;
+      }
+      long start = EnvironmentEdgeManager.currentTimeMillis();
+      try {
+          preBatchMutateWithExceptions(c, miniBatchOp);
+          return;
+      } catch (Throwable t) {
+          rethrowIndexingException(t);
+      } finally {
+          long duration = EnvironmentEdgeManager.currentTimeMillis() - start;
+          if (duration >= slowIndexPrepareThreshold) {
+              if (LOG.isDebugEnabled()) {
+                  LOG.debug(getCallTooSlowMessage("preBatchMutate", duration, slowIndexPrepareThreshold));
+              }
+              metricSource.incrementNumSlowIndexPrepareCalls();
+          }
+          metricSource.updateIndexPrepareTime(duration);
+      }
+      throw new RuntimeException(
+        "Somehow didn't return an index update but also didn't propagate the failure to the client!");
+  }
+
+  private static void setTimeStamp(KeyValue kv, byte[] tsBytes) {
+      int tsOffset = kv.getTimestampOffset();
+      System.arraycopy(tsBytes, 0, kv.getBuffer(), tsOffset, Bytes.SIZEOF_LONG);
+  }
+  private long getMaxTimestamp(Mutation m) {
+      long maxTs = 0;
+      long ts = 0;
+      Iterator iterator = m.getFamilyCellMap().entrySet().iterator();
+      while (iterator.hasNext()) {
+          Map.Entry<byte[], List<Cell>> entry = (Map.Entry) iterator.next();
+          Iterator<Cell> cellIterator = entry.getValue().iterator();
+          while (cellIterator.hasNext()) {
+              Cell cell = cellIterator.next();
+              ts = cell.getTimestamp();
+              if (ts > maxTs) {
+                  maxTs = ts;
+              }
+          }
+      }
+      return maxTs;
+  }
+
+  private void ignoreAtomicOperations (MiniBatchOperationInProgress<Mutation> miniBatchOp) {
+      for (int i = 0; i < miniBatchOp.size(); i++) {
+          Mutation m = miniBatchOp.getOperation(i);
+          if (this.builder.isAtomicOp(m)) {
+              miniBatchOp.setOperationStatus(i, IGNORE);
+              continue;
+          }
+      }
+  }
+
+  private void lockRows(MiniBatchOperationInProgress<Mutation> miniBatchOp, BatchMutateContext context) throws IOException {
+      for (int i = 0; i < miniBatchOp.size(); i++) {
+          if (miniBatchOp.getOperationStatus(i) == IGNORE) {
+              continue;
+          }
+          Mutation m = miniBatchOp.getOperation(i);
+          if (this.builder.isEnabled(m)) {
+              context.rowLocks.add(lockManager.lockRow(m.getRow(), rowLockWaitDuration));
+          }
+      }
+  }
+
+  private void populatePendingRows(BatchMutateContext context, long now) {
+      synchronized (this) {
```

Comment:
pendingRows is a per indexer object structure. Yes, this structure is populated with the keys of the rows that have been locked already. However, there can be multiple threads adding row keys to this structures simultaneously. The synchronization is necessary even though threads will add disjoint sets of row keys since HashMap is not thread safe. I was not aware of ConcurrentHashMap. I will use it and thus eliminate the synchronize blocks.

Reply:
Done